### PR TITLE
i#1621 AArch64 cc opt: Port inscount and compiler_inscount to AArch64.

### PR DESCRIPTION
**Pull request by fhahn**
This patch ports the inscount and compiler_inscount test cases to
AArch64 and adds them to the cleancall-opt-1 test cases. It also moves
the dr_init and dr_exit definitions to cleancall-opt-shared.h.

### CODE REVIEW

**Comment by fhahn** [comment]
The tests fail with 32 bit builds. I think that's because the wrong index for the clean call argument is used. Any ideas what the correct offset should be? Unfortunately I do not have access to a X86 machine setup to create 32 bit builds currently.

**Comment by zhaoqin** [comment]
Looks like the clean call context switch inlined, it make sense as the saving gpr on x86_32 is a single instruction.

```
162/269 Test #161: code_api|client.cleancall-opt-1 ..................................***Failed  Required regular expression not found.Regex=[^INIT
Calling func modify_gprs\.\.\.
Called func modify_gprs\.
Calling func inscount\.\.\.
Called func inscount\.
Calling func compiler_inscount\.\.\.
Called func compiler_inscount\.
PASSED
$
]  0.44 sec
INIT
Calling func modify_gprs...
Called func modify_gprs.
Calling func inscount...
global_count not updated properly after inscount!
Clean call code for inscount:
  0xef75ca54 mov    %eax -> %fs:0x00[4byte]
  0xef75ca5a mov    %fs:0x10[4byte] -> %eax
  0xef75ca60 mov    %esp -> 0x0c(%eax)[4byte]
  0xef75ca63 mov    0x00000168(%eax)[4byte] -> %esp
  0xef75ca69 mov    %fs:0x00[4byte] -> %eax
  0xef75ca6f push   $0x00000000 %esp -> %esp 0xfffffffc(%esp)[4byte]
  0xef75ca74 pushf  %esp -> %esp 0xfffffffc(%esp)[4byte]
  0xef75ca75 pusha  %esp %eax %ebx %ecx %edx %ebp %esi %edi -> %esp 0xffffffe0(%esp)[32byte]
  0xef75ca76 push   $0x0000dead %esp -> %esp 0xfffffffc(%esp)[4byte]
  0xef75ca7b call   $0xef7e0080 %esp -> %esp 0xfffffffc(%esp)[4byte]
  0xef75ca80 lea    0x04(%esp) -> %esp
  0xef75ca84 popa   %esp (%esp)[32byte] -> %esp %eax %ebx %ecx %edx %ebp %esi %edi
  0xef75ca85 popf   %esp (%esp)[4byte] -> %esp
  0xef75ca86 lea    0x0000011c(%esp) -> %esp
  0xef75ca8d mov    %eax -> %fs:0x00[4byte]
  0xef75ca93 mov    %fs:0x10[4byte] -> %eax
  0xef75ca99 mov    0x0c(%eax)[4byte] -> %esp
  0xef75ca9c mov    %fs:0x00[4byte] -> %eax
Called func inscount.
Calling func compiler_inscount...
Called func compiler_inscount.
PASSED
```

**Comment by zhaoqin** [comment]
Seems the arg offset seems right, 3.
```
CLEANCALL: decoding callee starting at: 0x4a828080
  0x4a828080  55                   push   %ebp %esp -> %esp 0xfffffffc(%esp)[4byte]
  0x4a828081  8b ec                mov    %esp -> %ebp
  0x4a828083  8b 45 0c             mov    0x0c(%ebp)[4byte] -> %eax
  0x4a828086  01 05 7c 23 78 f7    add    %eax 0xf778237c[4byte] -> 0xf778237c[4byte]
  0x4a82808c  c9                   leave  %ebp %esp (%ebp)[4byte] -> %esp %ebp
  0x4a82808d  c3                   ret    %esp (%esp)[4byte] -> %esp
```
Actually I am a bit worried that some compiler may optimize away the xbp save/restore since it is not necessary.

File: suite/tests/client-interface/cleancall-opt-1.dll.c [hunk 1]
```diff
@@ -109,7 +86,18 @@ event_basic_block(void *dc, void *tag, instrlist_t *bb,
      * We use a workaround involving ADR. */
     IF_AARCH64(save_current_pc(dc, bb, entry, &cleancall_start_pc, before_label));
     PRE(bb, entry, before_label);
-    dr_insert_clean_call(dc, bb, entry, func_ptrs[i], false, 0);
+
+    switch (i) {
+    default:
```
Comment (fhahn) [inline suite/tests/client-interface/cleancall-opt-1.dll.c]: I copied the switch structure from https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/client-interface/inline.dll.c#L222

**Comment by fhahn** [comment]
Looks like Windows and Unix both use the same stack index for the first argument. I've updated codegen_opnd_arg1 to use an offset of 2 on both Unix and Windows (this accounts for the push in the prologue). I'll submit this PR once the checks come back green once again

File: suite/tests/client-interface/cleancall-opt-shared.h [hunk 2]
```diff
@@ -518,6 +544,7 @@ before_callee(app_pc func, const char *func_name)
                           DR_MEMPROT_EXEC|DR_MEMPROT_READ|DR_MEMPROT_WRITE);
     }
 
+#ifdef TEST_INLINE
     ilist = instrlist_create(dc);
 #ifdef X86
```
Comment (zhaoqin) [inline suite/tests/client-interface/cleancall-opt-shared.h]: nit, we should adjust the indentation, also, the comment on #endif about which macro it ends would be helpful.

File: suite/tests/client-interface/cleancall-opt-shared.h [hunk 3]
```diff
@@ -606,3 +633,93 @@ codegen_empty(void *dc)
...
+static opnd_t
+codegen_opnd_arg1(void)
+{
+    /* FIXME: Perhaps DR should expose this.  It currently tracks this in
+     * core/instr.h. */
+#ifdef AARCH64
+  return opnd_create_reg(DR_REG_X0);
+#elif defined(X64)
+# ifdef UNIX
+    int reg = DR_REG_RDI;
```
Comment (zhaoqin) [inline suite/tests/client-interface/cleancall-opt-shared.h]: nit, an IF_WINDOWS_ELSE or something would be shorter.

File: suite/tests/client-interface/cleancall-opt-1.dll.c [hunk 4]
```diff
@@ -110,12 +87,23 @@ event_basic_block(void *dc, void *tag, instrlist_t *bb,
      * We use a workaround involving ADR. */
     IF_AARCH64(save_current_pc(dc, bb, entry, &cleancall_start_pc, before_label));
     PRE(bb, entry, before_label);
-   dr_insert_clean_call(dc, bb, entry, func_ptrs[i], false, 0);
+
+    switch (i) {
+    default:
```
Comment (zhaoqin) [inline suite/tests/client-interface/cleancall-opt-1.dll.c]: any reason the default is at the first?

Comment (fhahn) [inline suite/tests/client-interface/cleancall-opt-1.dll.c]: I copied the switch structure from https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/client-interface/inline.dll.c#L222

**Pull request merged**